dialogue-core/src/main/java/com/palantir/dialogue/core/DialogueConfig.java

```diff
@@ -0,0 +1,107 @@
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.core;
+
+import com.palantir.conjure.java.api.config.service.UserAgent;
+import com.palantir.conjure.java.client.config.ClientConfiguration;
+import com.palantir.logsafe.Preconditions;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.exceptions.SafeRuntimeException;
+import java.util.List;
+
+/**
+ * Configuration specifying everything necessary to talk to an abstract upstream 'service', with a number of
+ * possible uris. Intended to give us the flexibility to not require the legacy conjure-java-runtime jars at some
+ * point in the future.
+ *
+ * All getters package private initially.
+ */
+@SuppressWarnings("VisibilityModifier")
+public final class DialogueConfig {
+
+    // TODO(dfox): switch to getters when we can be bothered
+    public final ClientConfiguration legacyClientConfiguration;
+    final Class<? extends HttpChannelFactory> httpChannelFactory;
+    final UserAgent userAgent;
+
+    private DialogueConfig(Builder builder) {
+        this.legacyClientConfiguration =
+                Preconditions.checkNotNull(builder.legacyClientConfiguration, "legacyClientConfiguration");
+        this.httpChannelFactory = Preconditions.checkNotNull(builder.httpClientType, "httpClientType");
+        this.userAgent = Preconditions.checkNotNull(builder.userAgent, "userAgent");
+    }
+
+    List<String> uris() {
+        return legacyClientConfiguration.uris();
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static final class Builder {
+
+        private ClientConfiguration legacyClientConfiguration;
+        private Class<? extends HttpChannelFactory> httpClientType;
+        private UserAgent userAgent;
+
+        /** this method exists to provide a seamless transition from conjure-java-runtime. */
+        public Builder from(ClientConfiguration cjrClientConfig) {
+            this.legacyClientConfiguration = cjrClientConfig;
+            return this;
+        }
+
+        public Builder httpClientType(HttpClientType rawType) {
+            this.httpClientType = rawType.getFactoryClass();
+            return this;
+        }
+
+        public Builder userAgent(UserAgent value) {
+            this.userAgent = value;
+            return this;
+        }
+
+        public DialogueConfig build() {
+            return new DialogueConfig(this);
+        }
+    }
+
+    public enum HttpClientType {
+        APACHE("com.palantir.dialogue.hc4.DialogueApacheHttpClient"),
+        OKHTTP("TODO"),
+        HTTP_URL_CONNECTION("TODO"),
+        JAVA9_HTTPCLIENT("TODO");
+
+        private final String className;
+
+        HttpClientType(String className) {
+            this.className = className;
+        }
+
+        Class<? extends HttpChannelFactory> getFactoryClass() {
+            try {
+                return (Class<? extends HttpChannelFactory>) Class.forName(className);
```

Comment:
why do we need reflection here?

Reply:
i.e., why not `Supplier<Channel>`?

Reply:
I didn't want people who depend on dialogue-core to transitively get dialogue-apache, dialogue-java, dialogue-okhttp etc.

Might be able to doing this with a compileOnly dependency if you'd prefer?

Reply:
I'm not a huge fan of runtime dependencies like this. might indicate that the decomp is a little off, and maybe also that this should live in the server repo?